machine-learning/03-Training and AutoML.ipynb

```diff
@@ -1,45 +1,561 @@
+            ]
+          },
+          "metadata": {},
+          "output_type": "display_data"
+        }
+      ],
+      "source": [
+        "var rand = new Random(0);\n",
+        "var context =new MLContext(seed: 1);\n",
+        "var x = Enumerable.Range(-10000, 10000).Select(_x => _x * 0.1f).ToArray();\n",
+        "var y = x.Select(_x => 100 * _x + (rand.NextSingle() - 0.5f) * 10).ToArray();\n",
+        "var df = new DataFrame();\n",
+        "df[\"X\"] = DataFrameColumn.Create(\"X\", x);\n",
+        "df[\"y\"] = DataFrameColumn.Create(\"y\", y);\n",
+        "var trainTestSplit = context.Data.TrainTestSplit(df);\n",
+        "df.Head(10)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## Construct pipeline\n",
+        "The code below shows how to construct pipelines for both `SDCA` and `LightGbm`. The `Concatenate` transformer is necessary because it transfer a `single` column into `Vector<single>` type, which is the accepted feature type for both `SDCA` and `LightGbm` regressor."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "dotnet_interactive": {
+          "language": "csharp"
+        },
+        "vscode": {
+          "languageId": "dotnet-interactive.csharp"
+        }
+      },
+      "outputs": [],
+      "source": [
+        "var sdcaPipeline = context.Transforms.Concatenate(\"Features\", \"X\")\n",
+        "                            .Append(context.Regression.Trainers.Sdca(\"y\"));\n",
+        "var lgbmPipeline = context.Transforms.Concatenate(\"Features\", \"X\")\n",
+        "                            .Append(context.Regression.Trainers.LightGbm(\"y\"));"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## Train and evaluate model\n",
+        "The code below first trains `sdcaPipeline` and `lgbmPipeline` which are created above, then evaluate their performance on test dataset by calcuate `Root Mean Square Loss` between predicted and truth value. We can see that `SDCA` has better performance with a significant lower `Root Mean Square Loss` comparing with `LightGbm`, even it's a simple, linear model. This is because the training dataset is also linear, so `SDCA` can fit the dataset better than `LightGbm`."
```

Comment:
```suggestion
        "The code below first trains `sdcaPipeline` and `lgbmPipeline` which are created above, then evaluate their performance on test dataset by calculating `Root Mean Square Loss` between predicted and actual value. `SDCA` has better performance with a significantly lower `Root Mean Square Loss` compared to `LightGbm` even though it's a simpler linear model. This is because the training dataset is also linear, so `SDCA` can fit the dataset better than `LightGbm`."
```